### ISSUE TITLE
Bug Report: Application stop (& server shutdown) hangs due to HazelcastCachingProvider

### ISSUE THREAD

**Issue by meliora** [payara/Payara#5476]
# Description #
----------

In our app, we are using JavaMelody for monitoring. It seems that on shutdown, JavaMelody tries to write a monitoring summary to a file. This summary includes a summary of JCache and as this happens, a HazelCastClientCachingProvider gets instantiated which hangs the server.

## Expected Outcome ##

The server should shut down normally and most likely not create a new Hazelcast instance for JCache as it should be already present.

## Current Outcome ##

When we stop the application (deploy a new one, stop the server, undeploy the app ...) the server becomes unresponsive. We actually have to 'kill -9' the process to get it stopped. The server.log repeatedly prints out

```
[2021-10-28T13:21:33.063+0300] [Payara 5.2021.8] [WARNING] [] [com.hazelcast.client.impl.connection.ClientConnectionManager] [tid: _ThreadID=98 _ThreadName=admin-thread-pool::admin-listener(1)] [timeMillis: 1635416493063] [levelValue: 900] [[
  _hzinstance_jcache_shared [dev] [4.2] Unable to get live cluster connection, retry in 1473 ms, attempt: 9, cluster connect timeout: 9223372036854775807 ms, max backoff: 30000 ms]]
[2021-10-28T13:21:34.543+0300] [Payara 5.2021.8] [WARNING] [] [com.hazelcast.client.impl.connection.ClientConnectionManager] [tid: _ThreadID=98 _ThreadName=admin-thread-pool::admin-listener(1)] [timeMillis: 1635416494543] [levelValue: 900] [[
  _hzinstance_jcache_shared [dev] [4.2] Exception during initial connection to [127.0.0.1]:5701: com.hazelcast.core.HazelcastException: java.net.SocketException: Connection refused to address /127.0.0.1:5701]]
[2021-10-28T13:21:34.544+0300] [Payara 5.2021.8] [WARNING] [] [com.hazelcast.client.impl.connection.ClientConnectionManager] [tid: _ThreadID=98 _ThreadName=admin-thread-pool::admin-listener(1)] [timeMillis: 1635416494544] [levelValue: 900] [[
  _hzinstance_jcache_shared [dev] [4.2] Exception during initial connection to [127.0.0.1]:5702: com.hazelcast.core.HazelcastException: java.net.SocketException: Connection refused to address /127.0.0.1:5702]]
[2021-10-28T13:21:34.545+0300] [Payara 5.2021.8] [WARNING] [] [com.hazelcast.client.impl.connection.ClientConnectionManager] [tid: _ThreadID=98 _ThreadName=admin-thread-pool::admin-listener(1)] [timeMillis: 1635416494545] [levelValue: 900] [[
  _hzinstance_jcache_shared [dev] [4.2] Exception during initial connection to [127.0.0.1]:5703: com.hazelcast.core.HazelcastException: java.net.SocketException: Connection refused to address /127.0.0.1:5703]]
```

We took some thread dumps and traced the problem to JavaMelody:

```
"admin-thread-pool::admin-listener(1)" #50 daemon prio=5 os_prio=31 tid=0x00007fd87de95800 nid=0x14a03 waiting on condition [0x0000700004e17000]
   java.lang.Thread.State: TIMED_WAITING (sleeping)
	at java.lang.Thread.sleep(Native Method)
	at com.hazelcast.client.impl.connection.tcp.WaitStrategy.sleep(WaitStrategy.java:75)
	at com.hazelcast.client.impl.connection.tcp.TcpClientConnectionManager.doConnectToCandidateCluster(TcpClientConnectionManager.java:501)
	at com.hazelcast.client.impl.connection.tcp.TcpClientConnectionManager.doConnectToCluster(TcpClientConnectionManager.java:406)
	at com.hazelcast.client.impl.connection.tcp.TcpClientConnectionManager.connectToCluster(TcpClientConnectionManager.java:367)
	at com.hazelcast.client.impl.connection.tcp.TcpClientConnectionManager.start(TcpClientConnectionManager.java:316)
	- locked <0x0000000758780910> (a com.hazelcast.client.impl.connection.tcp.TcpClientConnectionManager)
	at com.hazelcast.client.impl.clientside.HazelcastClientInstanceImpl.start(HazelcastClientInstanceImpl.java:369)
	at com.hazelcast.client.HazelcastClient.constructHazelcastClient(HazelcastClient.java:460)
	at com.hazelcast.client.HazelcastClient.newHazelcastClientInternal(HazelcastClient.java:416)
	at com.hazelcast.client.HazelcastClient.newHazelcastClient(HazelcastClient.java:136)
	at com.hazelcast.client.cache.impl.HazelcastClientCachingProvider.getOrCreateInstanceByConfig(HazelcastClientCachingProvider.java:145)
	at com.hazelcast.client.cache.impl.HazelcastClientCachingProvider.getDefaultInstance(HazelcastClientCachingProvider.java:91)
	at com.hazelcast.cache.impl.AbstractHazelcastCachingProvider.getOrCreateInstance(AbstractHazelcastCachingProvider.java:341)
	at com.hazelcast.cache.impl.AbstractHazelcastCachingProvider.createHazelcastCacheManager(AbstractHazelcastCachingProvider.java:247)
	at com.hazelcast.cache.impl.AbstractHazelcastCachingProvider.getCacheManager(AbstractHazelcastCachingProvider.java:131)
	- locked <0x0000000758780c88> (a java.util.WeakHashMap)
	at com.hazelcast.cache.impl.AbstractHazelcastCachingProvider.getCacheManager(AbstractHazelcastCachingProvider.java:163)
	at com.hazelcast.cache.HazelcastCachingProvider.getCacheManager(HazelcastCachingProvider.java:275)
	at net.bull.javamelody.internal.model.JCacheInformations.buildJCacheInformationsList(JCacheInformations.java:95)
	at net.bull.javamelody.internal.model.JavaInformations.<init>(JavaInformations.java:204)
	at net.bull.javamelody.internal.web.HtmlController.writeHtmlToLastShutdownFile(HtmlController.java:384)
	at net.bull.javamelody.internal.web.MonitoringController.writeHtmlToLastShutdownFile(MonitoringController.java:291)
...
```

In HazelcastCachingProvider's javadoc, it says that if Hazelcast is the only caching provider present in the VM it initializes a new one when accessed via the cache-api. There is a similar issue in JavaMelody's issue database marked as won't fix: https://github.com/javamelody/javamelody/issues/858

### Workaround ###

We worked around the issue by adding 

 -Dhazelcast.jcache.provider.type=member

.. as a system property. This has an effect on HazelcastCachingProvider which does not instantiate the client-based implementation anymore and fixes the hang. As we are not experts of JCache, Hazelcast or its integration to Payara, we don't know if this has any side effects.  

## Steps to reproduce ## 

Deploy and undeploy an application with JavaMelody (1.88.0) included or maybe just access the default CachingManager via Caching.getCachingProvider().

## Environment ##

- **Distribution**: Payara 5.2021.8 Full Profile
- **JDK Version**: OpenJDK Runtime Environment (Zulu 8.58.0.13-CA-macosx)
- **Operating System**: Mac OSX

**Comment by MeroRai** [comment]
Please, can you provide a simple to follow scenario on how to reproduce this on the latest release of Payara Community Edition? A reproducer should ideally follow the SSCCE rules: http://www.sscce.org/. It will greatly help us to find the cause and fix it.

**Comment by meliora** [comment]
See https://github.com/meliora/payara-issue-5476

**Comment by meliora** [comment]
> Please, can you provide a simple to follow scenario on how to reproduce this on the latest release of Payara Community Edition? A reproducer should ideally follow the SSCCE rules: http://www.sscce.org/. It will greatly help us to find the cause and fix it.

Hi again, have you had time to reproduce this issue? It should be easy enough with the code/ear provided.

**Comment by MeroRai** [comment]
Hi @meliora, 

Apologise for the late reply. Payara Server already comes with quite comprehensive tools to monitor the application, which we would recommend you use instead of JavaMelody. You can read more about these monitoring tools in our documentation using the following links:

- https://docs.payara.fish/community/docs/documentation/user-guides/monitoring/monitoring-console.html
- https://docs.payara.fish/community/docs/documentation/user-guides/monitoring/monitoring.html
- https://docs.payara.fish/community/docs/documentation/payara-server/monitoring-service/monitoring-service.html

**Comment by meliora** [comment]
@MeroRai The problem is not JavaMelody. Problem is that when using JSR107 JCache API, Payara breaks down as shown by the reproducer you asked.

Have you looked at the reproducer we provided? It does not use JavaMelody - it just uses JSR107 (which JavaMelody does too - incidentally).

Just to note: JavaMelody is a quite widely used library for monitoring. We have a lot of things depending on it. Replacing it with other tools is not an option.

**Comment by tomekkup** [comment]
i have the same problem which is actually not allowing me to use javamelody on payara 5.2022
